miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,677 @@
+    pub fn depth(&self) -> u8 {
+        (self.nodes.len() + self.empty_nodes_mask.count_ones() as usize) as u8
+    }
+
+    /// Get a specific node in this path at a given depth.
+    ///
+    /// The `depth` parameter is defined in terms of `self.depth()`. Merkle paths conventionally do
+    /// not include the root, so the shallowest depth is `1`, and the deepest depth is
+    /// `self.depth()`.
+    ///
+    /// # Errors
+    /// Returns [MerkleError::DepthTooBig] if `node_depth` is greater than the total depth of this
+    /// path.
+    pub fn at_depth(&self, node_depth: NonZero<u8>) -> Result<RpoDigest, MerkleError> {
+        if node_depth.get() > self.depth() {
+            return Err(MerkleError::DepthTooBig(node_depth.get().into()));
+        }
+
+        let node = self
+            .at_depth_nonempty(node_depth)
+            .unwrap_or_else(|| *EmptySubtreeRoots::entry(self.depth(), node_depth.get()));
+
+        Ok(node)
+    }
+
+    /// Get a specific non-empty node in this path at a given depth, or `None` if the specified
+    /// node is an empty node.
+    fn at_depth_nonempty(&self, node_depth: NonZero<u8>) -> Option<RpoDigest> {
+        assert!(
+            node_depth.get() <= self.depth(),
+            "node depth {} cannot be greater than tree depth {}",
+            node_depth.get(),
+            self.depth(),
+        );
+
+        if self.is_depth_empty(node_depth) {
+            return None;
+        }
+
+        // Our index needs to account for all the empty nodes that aren't in `self.nodes`.
+        let nonempty_index = self.get_nonempty_index(node_depth);
+
+        Some(self.nodes[nonempty_index])
+    }
+
+    /// Returns the path node at the specified index, or [None] if the index is out of bounds.
+    ///
+    /// The node at index 0 is the deepest part of the path.
+    ///
+    /// ```
+    /// # use core::num::NonZero;
+    /// # use miden_crypto::{ZERO, ONE, hash::rpo::RpoDigest, merkle::SparseMerklePath};
+    /// # let zero = RpoDigest::new([ZERO; 4]);
+    /// # let one = RpoDigest::new([ONE; 4]);
+    /// # let sparse_path = SparseMerklePath::from_sized_iter(vec![zero, one, one, zero]).unwrap();
+    /// let depth = NonZero::new(sparse_path.depth()).unwrap();
+    /// assert_eq!(
+    ///     sparse_path.at_idx(0).unwrap(),
+    ///     sparse_path.at_depth(depth).unwrap(),
+    /// );
+    /// ```
+    pub fn at_idx(&self, index: usize) -> Option<RpoDigest> {
+        // If this overflows *or* if the depth is zero then the index was out of bounds.
+        let depth = NonZero::new(u8::checked_sub(self.depth(), index as u8)?)?;
+        self.at_depth(depth).ok()
+    }
+
+    // PROVIDERS
+    // ============================================================================================
+
+    /// Constructs a borrowing iterator over the nodes in this path.
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = RpoDigest> {
```

Comment:
Could we specify the direction of iteration here? i.e., is it from the root to the leaf or from the leaf to the root?

Reply:
Added a comment both here and on the `SparseMerklePathIter` (it's a public type).